test/FunctionalTests/Microsoft.OData.Client.Tests/ALinq/SequenceMethodsTests.cs

```diff
@@ -0,0 +1,336 @@
+            Assert.Throws<InvalidOperationException>(() => _ctx.Customers.Single(c => c.Name == "thisdoesntexist"));
+        }
+
+        [Fact]
+        public void SinglePredicate_ThrowsException_WhenMoreThanOneMatchExists()
+        {
+            _ctx.OnRequestUriBuilt = (string builtUri) =>
+            {
+                string expectedUri = _ctx.BuildUriFromPath("/Customers?$filter=ContactName ne 'thisdoesntexist'&$top=2");
+                Assert.Equal(expectedUri, builtUri);
+            };
+            _ctx.InterceptRequestAndMockResponseValue("Customers", "[" + TestContext.MockCustomer1 + "," + TestContext.MockCustomer2 + "]");
+            Assert.Throws<InvalidOperationException>(() => _ctx.Customers.Single(c => c.Name != "thisdoesntexist"));
+        }
+
+        [Fact]
+        public void SingleOrDefaultPredicate()
+        {
+            _ctx.OnRequestUriBuilt = (string builtUri) =>
+            {
+                string expectedUri = _ctx.BuildUriFromPath("/Customers?$filter=CustomerID eq 'CHOPS'&$top=2");
+                Assert.Equal(expectedUri, builtUri);
+            };
+            _ctx.InterceptRequestAndMockResponseValue("Customers", "[" + TestContext.MockCustomer2 + "]");
+            Customer customer = _ctx.Customers.SingleOrDefault(c => c.Id == "CHOPS");
+            Assert.Equal("CHOPS", customer.Id);
+            Assert.Equal("Yang Wang", customer.Name);
+            Assert.Equal("Bern", customer.City);
+        }
+
+        [Fact]
+        public void SingleOrDefaultPredicate_ReturnsNull_WhenNoMatchExists()
+        {
+            _ctx.OnRequestUriBuilt = (string builtUri) =>
+            {
+                string expectedUri = _ctx.BuildUriFromPath("/Customers?$filter=CustomerID eq '234111'&$top=2");
+                Assert.Equal(expectedUri, builtUri);
+            };
+            _ctx.InterceptRequestAndMockResponseValue("Customers", "[]");
+            Assert.Null(_ctx.Customers.SingleOrDefault(c => c.Id == "234111"));
+        }
+
+        [Fact]
+        public void SingleOrDefaultPredicate_ThrowsException_WhenMoreThanOneMatchExists()
+        {
+            _ctx.OnRequestUriBuilt = (string builtUri) =>
+            {
+                string expectedUri = _ctx.BuildUriFromPath("/Customers?$filter=ContactName ne 'thisdoesntexist'&$top=2");
+                Assert.Equal(expectedUri, builtUri);
+            };
+            _ctx.InterceptRequestAndMockResponseValue("Customers", "[" + TestContext.MockCustomer1 + "," + TestContext.MockCustomer2 + "]");
+            Assert.Throws<InvalidOperationException>(() => _ctx.Customers.SingleOrDefault(c => c.Name != "thisdoesntexist"));
+        }
+    }
+
+
+    [Key("CustomerID")]
+    public class Customer
+    {
+        [OriginalName("CustomerID")]
+        public string Id { get; set; }
+
+        public string City { get; set; }
+
+        [OriginalName("ContactName")]
+        public string Name { get; set; }
+    }
+
+    public class TestContext
```

Comment:
While there are many ways of achieving the same goal, I can't quite understand what value this `TestContext` class brings. In my opinion, it has made the logic more complex that it needs to be - for example having to introduce `Action<string> OnRequestUriBuilt` delegate field to fulfill the requirement to validate the request Uri. It does help to familiarize with how things are done elsewhere in the codebase, otherwise we end up with code that is not easy to follow or maintain. Even where a custom "context" class is needed, the common practice is to subclass `DataServiceContext` instead of composition like you're doing here.

I understand what you are trying to achieve by introducing the `TestContext` but the same could easily be achieved differently, easily, neatly, and with less complexity...
E.g.
```csharp
// From the test class constructor
_ctx = new DataServiceContext();
var model = BuildEdmModel();
_ctx.Format.UseJson(model);
_ctx.ResolveName = (type) => $"NS.{type.Name}";
_ctx.KeyComparisonGeneratesFilterQuery = true;

// ... other code

// From test method
// DO: Call InterceptRequestAndMockResponse
// DO: _ctx.OnBuildingRequest += ...
Customer customer _ctx.CreateQuery<Customer>("Customers").FirstOrDefault(c => c.Name == "John")

// DO: Assert statements
```

**No need to use TestContext anywhere**

Please take a look at this for reference: https://github.com/OData/odata.net/blob/383e29f3cb96eda22e76d19df7bdab11459377b8/test/FunctionalTests/Microsoft.OData.Client.Tests/ALinq/DollarApplyTestsBase.cs#L27-L42

Reply:
I wanted to clarify the reasoning behind the implementation of the TestContext class and address some of the concerns raised.

I implemented the TestContext class to separate the logic of the test methods from the connection logic to the data service. This separation aims to bring more cleanliness and organization to the code, contrary to the opinion expressed. The goal is to keep the test methods focused solely on testing, while the mock connection and setup logic is handled elsewhere.

The logic in this file is extremely simple and should not pose a problem for any maintainer. In fact, when comparing the complexity of this file to the overall project (OData), the complexity here is minimal. I find the continuous requests for changes a bit excessive and not entirely justified given the simplicity of the implementation.

Regarding Inheritance, I agree with the point that TestContext could extend DataServiceContext. However, I did not implement it this way initially because I adapted this class from another file where I needed to use four different contexts. I can modify this, but I sincerely believe it is not necessary.

I would appreciate it if we could focus on impactful change requests rather than minor issues. If there are substantial improvements to be made, I am more than willing to address them, but I feel that some of the requests so far have been less about meaningful improvements and more about nitpicking.

Thank you for your understanding.

Reply:
Let me contribute my thoughts to this. While there have been a lot of change requests to what is seemingly a simple change, I would argue that those changes have meaningfully improved the completeness and readability of the tests. More edge cases are handled, and tests are easier to read now than in the initial PR.

On this particular issue of the `TestContext`, I think the helper methods like `InterceptRequestAndMockResponse` serve to declutter the test code from boilerplate and noise. However, I don't think having such helper methods defined in the test class, and initial setup code in the test class constructor would be any less readable than having them in a new `TestContext` class. We can easily distinguish private helper methods from test methods with proper organization in the class and test attributes (`[Fact]`, `[Theory]`). For that reason, I think having a separate `TestContext` class that's only used one is unnecessary. It would see the benefit if the helpers were re-used in multiple test classes, but that's not the case here. I would argue that having classes we don't need does impact maintainability because it increases the surface areas of things we're potentially concerned about and things that would cause us to ask "So, what do you really do here?" or wonder whether it's safe to change. The less such things we have, the better.

That said, conceding that my view on this particular change is subjective, I don't think it's worthwhile to hold the change back from going in.

Reply:
Unfortunately, I have changed my view on this matter. Having gone through this other PR: https://github.com/OData/odata.net/pull/2862 which has a similar pattern, I'm concerned about letting this in as is and then resulting in a proliferation of `TestContext` pattern in future tests. I think it would be easier to refactor it at this point.